listeners.go

```diff
@@ -453,12 +453,18 @@ func ListenQUIC(ln net.PacketConn, tlsConf *tls.Config, activeRequests *int64) (
 		if err != nil {
 			return nil, err
 		}
-		return &sharedQuicListener{EarlyListener: earlyLn, key: lnKey}, nil
+		return &sharedQuicListener{EarlyListener: earlyLn, tlsConf: tlsConf, key: lnKey}, nil
 	})
 	if err != nil {
 		return nil, err
 	}
 
+	sql := sharedEarlyListener.(*sharedQuicListener)
+	// update GetConfigForClient field because both quic and caddy tls uses this field
+	if sql.tlsConf != tlsConf {
+		sql.tlsConf.GetConfigForClient = tlsConf.GetConfigForClient
```

Comment:
Huh -- interesting! Is this thread-safe? (Run with race detector)

Reply:
Is `XCADDY_RACE_DETECTOR=1` when building enough?

Also this approach suffers from config rollback. I prefer just creating new quic listeners.

Reply:
Re-creating the listeners means downtime, doesn't it? Reloads are meant to be graceful, that's why listeners are persisted across config reloads.

Reply:
I don't know the details of quic listeners. They share the underlying udp sockets from the first glance.

Reply:
@WeidiDeng 

> Is XCADDY_RACE_DETECTOR=1 when building enough?

Should be -- just be sure the relevant code path gets run with that variable set. (You should notice the program be slower when the race detector is enabled.)

Reply:
I'm thinking about creating a new struct to delegate `GetConfigForClient` to. It will hopefully deal with config rollback as @francislavoie  mentioned in slack.

Reply:
I guess I don't actually understand why this is the problem:

> because caddy tries to reuse existing quic listeners.

We reuse the listeners for normal HTTP too, and those work fine. What exactly about the code is causing the old certificates to persist even after they are renewed?

Reply:
The problem is where the tls.Config is from `tls app`. Because of listener pool, shared quic listener is used between config reload. However, the old tls.Config is still in use, whose GetCertificate is still used by quic early listener.

Normal http don't bind tls.Config to the lifetime of the listener. I believe every time tls.Newlistener is called in the `http start` method.

Reply:
I'm currently working on it. I'm wondering if it's possible for ListenQuic to be called using different tls.Config. Normally it's called by starting and reloading caddy. At most there should only be 2 instances of tls.Config for a single quic listener?

Reply:
(Oh, duh, now I see that you already mentioned this same solution above. :man_facepalming: )

> At most there should only be 2 instances of tls.Config for a single quic listener?

Wouldn't there only be 1 per listener? (Sorry, not sure I understand.)

Reply:
I mean different call paths. There is only one tls.Confif per listener, but underlying GetConfigForClient can be more than 1. I want to make sure that theoretically, underlying GetConfig can at most be 2, one from old config, one from new config.

Reply:
Yeah at most two _per listen address_. Users might use quic on other ports than 443